### PR TITLE
feat: allow Ns in the sample barcode

### PR DESCRIPTION
**Pull request by nh13**
This is also useful if the barcodes are of different length, so we can pad the shorter barcodes with Ns

#29

### CODE REVIEW

File: src/lib/barcode_matching.rs [hunk 1]
```diff
@@ -1,14 +1,10 @@
+use super::byte_is_nocall;
 use ahash::HashMap as AHashMap;
 use ahash::HashMapExt;
 use bstr::{BString, ByteSlice};
 
 const STARTING_CACHE_SIZE: usize = 1_000_000;
 
-/// Checks whether a given u8 byte is a "No-call"-ed base, signified by the bytes 'N', 'n' and '.'
```
Comment (nh13) [inline src/lib/barcode_matching.rs:7]: lots of changes since I pulled two methods out into mod.rs

File: src/lib/samples.rs [hunk 2]
```diff
@@ -207,6 +203,14 @@ mod tests {
         assert!(samples_metadata.samples[1].barcode == "CATGCTA");
     }
 
+    #[test]
+    fn test_new_sample_non_agct_bases_in_barcode() {
```
Comment (tfenne) [inline src/lib/samples.rs:207]:
```suggestion
    fn test_new_sample_non_agct_bases_in_barcode_allowed() {
```

**Review by tfenne — APPROVED** [review]
Could you please also add to the usage for the demux command a paragraph explaining how Ns in the expected barcodes are handled and how what is and isn't counted as mismatching?

File: src/bin/commands/demux.rs [hunk 3]
```diff
@@ -466,6 +466,15 @@ impl DemuxMetric {
 /// molecular identifiers from each read.  The observed sample barcode will be matched to the
 /// sample barcodes extracted from the bases in the sample metadata and associated read structures.
 ///
+/// An observed barcode matches an expected barcocde if all the following are true:
+/// 1. The number of mismatches (edits/substitutions) is less than or equal to the maximum 
+///.   mismatches (see `--max-mismatches`).
```
Comment (nh13) [inline src/bin/commands/demux.rs:471]:
```suggestion
///    mismatches (see `--max-mismatches`).
```

**Pull request merged**